File: languages/go/uhppote/decode.go
package uhppote

import (
    "encoding/binary"
    "encoding/hex"
    "fmt"
    "net/netip"
    "time"
)

{{range .model.responses}}{{template "decode" .}}
{{end}}

{{with .model.event}}{{template "decode" .}}
{{end}}

{{range .model.responses}}{{template "response" .}}
{{end}}

{{with .model.event}}
type {{CamelCase .name}} struct { {{range .fields}}
    {{CamelCase .name}} {{template "type" .type}} `json:"{{kebabCase .name}}"`{{end}}
}
{{end}}

{{define "decode"}}
func {{camelCase .name}}(packet []byte) (*{{CamelCase .name}}, error) {
    if len(packet) != 64 {
        return nil, fmt.Errorf("invalid reply packet length (%v)", len(packet))
    }

    if packet[1] != {{byte2hex .msgtype}} {
        return nil, fmt.Errorf("invalid reply function code (%02x)", packet[1])
    }

    response := {{CamelCase .name}}{}
    {{range .fields}}
    if v,err := unpack{{CamelCase .type}}(packet, {{.offset}}); err != nil {
        return nil, err
    } else {
        response.{{CamelCase .name}} = v
    }
    {{end}}

    return &response, nil
}
{{end}}

func unpackUint8(packet []byte, offset uint8) (uint8, error) {
    return packet[offset], nil
}

func unpackUint16(packet []byte, offset uint8) (uint16, error) {
    return binary.LittleEndian.Uint16(packet[offset:offset+2]), nil
}

func unpackUint32(packet []byte, offset uint8) (uint32, error) {
    return binary.LittleEndian.Uint32(packet[offset:offset+4]), nil
}

func unpackBool(packet []byte, offset uint8) (bool, error) {
    return packet[offset] != 0x00, nil
}

func unpackIPv4(packet []byte, offset uint8) (netip.Addr, error) {
    addr,_ := netip.AddrFromSlice(packet[offset:offset+4])

    return addr, nil
}

func unpackMAC(packet []byte, offset uint8) (string, error) {
    mac := fmt.Sprintf("%02x:%02x:%02x:%02x:%02x:%02x", 
        packet[offset],
        packet[offset+1],
        packet[offset+2],
        packet[offset+3],
        packet[offset+4],
        packet[offset+5])

    return mac, nil
}

func unpackVersion(packet []byte, offset uint8) (string, error) {
    major := packet[offset]
    minor := packet[offset+1]

    return fmt.Sprintf("v%x.%02x", major, minor), nil
}

func unpackDate(packet []byte, offset uint8) (Date, error) {
    bcd := bcd2string(packet[offset:offset+4])

    if date, err := time.ParseInLocation("20060102", bcd, time.Local); err != nil {
        return Date(time.Time{}), nil
    } else {
        return Date(date), nil
    }
}

func unpackShortdate(packet []byte, offset uint8) (Date, error) {
    bcd := "20" + bcd2string(packet[offset:offset+3])

    if date, err := time.ParseInLocation("20060102", bcd, time.Local); err != nil {
        return Date(time.Time{}), nil
    } else {
        return Date(date), nil
    }
}

func unpackTime(packet []byte, offset uint8) (Time, error) {
    bcd := bcd2string(packet[offset:offset+3])

    if t, err := time.ParseInLocation("150405", bcd, time.Local); err != nil {
        return Time(time.Time{}), nil
    } else {
        return Time(t), nil
    }
}

func unpackDatetime(packet []byte, offset uint8) (DateTime, error) {
    bcd := bcd2string(packet[offset:offset+7])

    if date, err := time.ParseInLocation("20060102150405", bcd, time.Local); err != nil {
        return DateTime(time.Time{}), nil
    } else {
        return DateTime(date), nil
    }
}

func unpackHHmm(packet []byte, offset uint8) (HHmm, error) {
    bcd := bcd2string(packet[offset:offset+2])

    if t, err := time.ParseInLocation("1504", bcd, time.Local); err != nil {
        return HHmm(time.Time{}), nil
    } else {
        return HHmm(t), nil
    }
}

func bcd2string(bytes []byte) string {
    BCD := hex.EncodeToString(bytes)

    if matched,err := strings.MatchString(`^[0-9]*$`, BCD); err != nil || !matched {
            panic(fmt.Sprintf("invalid BCD value (%v)", bytes))        
    } 

    return BCD
}

{{define "response"}}
type {{CamelCase .name}} struct { {{range .fields}}
    {{CamelCase .name}} {{template "type" .type}} `json:"{{kebabCase .name}}"`{{end}}
}
{{end}}

File: languages/http/decode.js
{{range .model.responses}}{{template "decode" .}}
{{end}}

{{define "decode"}}
export function {{CamelCase .name}}(packet) {
  const buffer = new Uint8Array(packet)
  const view = new DataView(buffer.buffer)

  if (buffer.length !== 64) {
    throw new Error(`invalid reply packet length (${buffer.length})`)
  }

  if (buffer[1] !== {{byte2hex .msgtype}}) {
    throw new Error(`invalid reply function code (${buffer[1].toString(16).padStart(2,'0')})`)
  }

  return { {{range .fields}}
    {{camelCase .name}}: unpack{{CamelCase .type}}(view, {{.offset}}),{{end}}
  }
}
{{end}}

function unpackUint8(packet, offset) {
  return packet.getUint8(offset)
}

function unpackUint16(packet, offset) {
  return packet.getUint16(offset, true)
}

function unpackUint32(packet, offset) {
  return packet.getUint32(offset, true)
}

function unpackBool(packet, offset) {
  return packet.getUint8(offset) !== 0x00
}

function unpackIPv4(packet, offset) {
  const bytes = new Uint8Array(packet.buffer.slice(offset, offset + 4))

  return [...bytes].map(x => x.toString(10)).join('.')
}

function unpackMAC(packet, offset) {
  const bytes = new Uint8Array(packet.buffer.slice(offset, offset + 6))
  
  return [...bytes].map(x => x.toString(16).padStart(2, '0')).join(':')
}

function unpackVersion(packet, offset) {
  const major = packet.getUint8(offset).toString(16)
  const minor = packet.getUint8(offset+1).toString(16).padStart(2, '0')

  return `v${major}.${minor}`
}

function unpackDate(packet, offset) {
  const bytes = new Uint8Array(packet.buffer.slice(offset, offset + 4))
  const datetime = bcd(bytes)

  if (datetime === '00000000') {
    return ''
  }

  const date = `${datetime.substr(0, 4)}-${datetime.substr(4, 2)}-${datetime.substr(6, 2)}`

  return `${date}`
}

function unpackShortdate(packet, offset) {
  const bytes = new Uint8Array(packet.buffer.slice(offset, offset + 3))
  const datetime = bcd('20' + bytes)

  if (datetime === '20000000') {
    return ''
  }

  const date = `${datetime.substr(0, 4)}-${datetime.substr(4, 2)}-${datetime.substr(6, 2)}`

  return `${date}`
}

function unpackDatetime(packet, offset) {
  const bytes = new Uint8Array(packet.buffer.slice(offset, offset + 7))
  const datetime = bcd(bytes)

  if (datetime === '00000000000000') {
    return ''
  }

  const date = `${datetime.substr(0, 4)}-${datetime.substr(4, 2)}-${datetime.substr(6, 2)}`
  const time = `${datetime.substr(8, 2)}:${datetime.substr(10, 2)}:${datetime.substr(12, 2)}`

  return `${date} ${time}`
}

function unpackTime(packet, offset) {
  const bytes = new Uint8Array(packet.buffer.slice(offset, offset + 3))
  const datetime = bcd(bytes)

  if (datetime === '000000') {
    return ''
  }

  const time = `${datetime.substr(8, 2)}:${datetime.substr(10, 2)}:${datetime.substr(12, 2)}`

  return `${time}`
}

function unpackHHmm(packet, offset) {
  const bytes = new Uint8Array(packet.buffer.slice(offset, offset + 2))
  const datetime = bcd(bytes)

  if (datetime === '0000') {
    return ''
  }

  const time = `${datetime.substr(8, 2)}:${datetime.substr(10, 2)}}`

  return `${time}`
}

function bcd (bytes) {
  return [...bytes].map(x => [(x >>> 4) & 0x0f, (x >>> 0) & 0x0f]).flat().join('')
}
